starter_kits/D/hlt/entity.d

```diff
@@ -0,0 +1,74 @@
+
+import constants;
+import positionals : Position, Direction;
+import types;
+
+import std.stdio : write;
+import std.format;
+
+abstract class Entity
+{
+  immutable EntityId id;
+  immutable PlayerId ownerId;
+  immutable Position position;
+
+  this(EntityId id, PlayerId ownerId, Position position)
+  {
+    this.id = id;
+    this.ownerId = ownerId;
+    this.position = position;
+  }
+}
+
+class Dropoff : Entity
+{
+  this(EntityId id, PlayerId ownerId, Position position)
+  {
+    super(id, ownerId, position);
+  }
+}
+
+class Shipyard : Entity
+{
+  this(PlayerId owner, int x, int y)
+  {
+    super(-1, owner, Position(x, y));
+  }
+
+  final void spawn() const
+  {
+    write("g");
```

Comment:
Yes, the server expects a single line per turn, which is the case here (`write` does not append a new line).

However, no, the server does not expect spaces: it parses the commands by first reading the command type as a single `char` [1], then reading the necessary parameters for that command. This works regardless of whether the commands were space separated, because the parser only ever reads numbers and single characters, so there is no ambiguity.

I don't mind adding spaces if you feel it's more clear though.

[1] https://github.com/HaliteChallenge/Halite-III/blob/master/game_engine/core/command/Command.cpp#L49

Reply:
I think you should double-check. Server really does expect one line per turn with space between commands. For example, how would it be able to tell when you are "done" sending commands for the turn?

Reply:
That's the reason for central "end turn" function too in other kits, so that it can send the final end-of-line character.

Reply:
Oh sorry, I misread.

Reply:
Yeah, I see now, I didn't notice the doneSending and I assumed write must be adding line-end itself.

Reply:
I would prefer that you add terminal space - I know of at least one alternative implementation of the engine and I think it's better not to assume that it does reads the same way.